### DISCUSSION TITLE
Issue when use typescript and styled-component

### DISCUSSION THREAD

**Discussion by meirbo** [vercel/next.js#35905] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 11
    Binaries:
      Node: 16.14.2
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.1.5-canary.1
      react: 17.0.2
      react-dom: 17.0.2

### Which example does this report relate to?

with-react-native-web

### Describe the Bug

Hello, 

I tried to implement with the example with-react-native-web, styled-component and typescript. 

I got many different issue. 

When i tried to use styled-components/native, i got this issue : error - SyntaxError: Cannot use import statement outside a module
When i tried to use styled-components, i got issue with withConfig undefined. 

Thanks for the help. 

### Expected Behavior

Styled-components working with typescript with nextjs and react-native-web

### To Reproduce

Use this repo, install typescript and config your appli with it. install styled-components and try to use it

**Comment by steve-taylor** [comment]
I have the same issue with Styled Components and the regular version of React. This happens whether I use SWC or Babel.

```
info  - Skipping linting
info  - Checking validity of types  
info  - Disabled SWC as replacement for Babel because of custom Babel configuration ".babelrc" https://nextjs.org/docs/messages/swc-disabled
info  - `compiler` options in `next.config.js` will be ignored while using Babel https://nextjs.org/docs/messages/ignored-compiler-options
info  - Using external babel configuration from /path/to/monorepo/apps/app/.babelrc
info  - Creating an optimized production build  
info  - Compiled successfully
info  - Collecting page data ...node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

TypeError: Cannot read properties of undefined (reading 'withConfig')
    at Object.6826 (/path/to/monorepo/apps/app/.next/server/chunks/4817.js:512:62)
    at __webpack_require__ (/path/to/monorepo/apps/app/.next/server/webpack-runtime.js:25:42)
    at Object.6476 (/path/to/monorepo/apps/app/.next/server/chunks/8448.js:51:72)
    at __webpack_require__ (/path/to/monorepo/apps/app/.next/server/webpack-runtime.js:25:42)
    at /path/to/monorepo/apps/app/.next/server/pages/_app.js:839:77
    at Function.__webpack_require__.a (/path/to/monorepo/apps/app/.next/server/webpack-runtime.js:91:13)
    at Object.4645 (/path/to/monorepo/apps/app/.next/server/pages/_app.js:834:21)
    at __webpack_require__ (/path/to/monorepo/apps/app/.next/server/webpack-runtime.js:25:42)
    at __webpack_exec__ (/path/to/monorepo/apps/app/.next/server/pages/_app.js:2326:39)
    at /path/to/monorepo/apps/app/.next/server/pages/_app.js:2327:114 {
  type: 'TypeError'
}
```

Not sure why this was moved from issue to discussion.

**Reply by icyJoseph** [reply]
Hi,

Do you have a reproduction demo? If you could use https://stackblitz.com/, or some other means to prepare one?

I ask this because I have styled components in a couple of sites with Next.js, both with Babel and SWC, and I see no issues at all.

**Reply by steve-taylor** [reply]
I don't have a reproduction yet. It will take some time to prepare. For now, here's a sample of the output chunk where the error occurs:

```js
var Svg = external_styled_components_.svg.withConfig({
  displayName: "Icon__Svg",
  componentId: "sc-19bftdh-0"
})(["display:inline-block;width:100%;height:100%;fill:currentColor;stroke:none;pointer-events:none;"]);
function Icon_Icon(_ref) {
  var viewBox = _ref.viewBox,
      ariaLabel = _ref.ariaLabel,
      children = _ref.children,
      _ref$ariaHidden = _ref.ariaHidden,
      ariaHidden = _ref$ariaHidden === void 0 ? ariaLabel ? undefined : true : _ref$ariaHidden,
      attrs = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/jsx_runtime_.jsx(Svg, _objectSpread(_objectSpread({}, attrs), {}, {
    viewBox: viewBox,
    "aria-hidden": ariaHidden,
    "aria-label": ariaLabel,
    children: children
  }));
}
```

The error occurs on the first line because `external_styled_components_.svg` is `undefined`.

In my other package's transpiled `Icon.tsx`, I hacked this in:

```
console.log(styled);
```

which revealed that the `styled` object contains the entire module instead of the default export. So this looks like a bug in `styled-components`. This happens regardless of whether or not the Next.js app's package.json has `"type": "module"`.

Interestingly, looking at `.next/server/pages/_app.js`, I see this:

```js
/* harmony import */ var styled_components__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(7518);
/* harmony import */ var styled_components__WEBPACK_IMPORTED_MODULE_2___default = /*#__PURE__*/__webpack_require__.n(styled_components__WEBPACK_IMPORTED_MODULE_2__);
```

whereas in the chunk shown above, I see this:

```js
// EXTERNAL MODULE: external "styled-components"
var external_styled_components_ = __webpack_require__(7518);
```

Notice the difference? The `_app.js` chunk treats `__webpack_require__(7518)` as the entire module, whereas `4817.js` incorrectly treats `__webpack_require__(7518)` as the default export. Perhaps `external_styled_components_` should use `__webpack_require__.n` instead of just `__webpack_require__`.

**Reply by icyJoseph** [reply]
Are you using any other libraries that use `styled-components`?

**Reply by steve-taylor** [reply]
My other packages in the same monorepo use `styled-components`. They’re ESM packages transpiled with Babel. Each of their package.json has `"type": "module"` and a `"main"` pointing to an `index.js` file that re-exports from other transpiled .js files. I don’t bundle those other packages because I don’t see the point of it when `next build` will bundle them anyway. (This means that every dependency has to be a peer dependency.)

**Reply by steve-taylor** [reply]
@icyJoseph here's the reproduction: https://github.com/steve-taylor/next-styled-esm-repro

**Reply by icyJoseph** [reply]
I couldn't get `bolt` to work, perhaps because of my OS, but I did manage to see the issue after some tweaking.

- https://github.com/styled-components/styled-components/issues/3601
- https://github.com/styled-components/styled-components/issues/3437

Doesn't it feel like this is the issue?

**Reply by steve-taylor** [reply]
Definitely the first issue. I commented there with my reproduction as someone else mentioned Next.js. Thanks for the links.

**Reply by tofuwu-dev** [reply]
Which compiler do you use in your project? babel or swc?

**Reply by steve-taylor** [reply]
I use SWC, but the issue occurs regardless of SWC or Babel.

**Comment by steve-taylor** [comment]
I have a workaround! See the [with-preconstruct-react-markdown-conditional-exports](https://github.com/steve-taylor/next-styled-esm-repro/tree/with-preconstruct-react-markdown-conditional-exports) branch. README.md explains the workaround.
